x-pack/elastic-agent/pkg/agent/transpiler/vars.go

```diff
@@ -0,0 +1,199 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package transpiler
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/elastic/go-ucfg"
+)
+
+var varsRegex = regexp.MustCompile(`\${([\p{L}\d\s\\\-_|.'"]*)}`)
+
+// ErrNoMatch is return when the replace didn't fail, just that no vars match to perform the replace.
+var ErrNoMatch = fmt.Errorf("no matching vars")
+
+// Vars is a context of variables that also contain a list of processors that go with the mapping.
+type Vars struct {
+	Mapping map[string]interface{}
+
+	ProcessorsKey string
+	Processors    Processors
+}
+
+// Replace returns a new value based on variable replacement.
+func (v *Vars) Replace(value string) (Node, error) {
+	var processors []map[string]interface{}
+	c, err := ucfg.NewFrom(v.Mapping, ucfg.PathSep("."))
+	if err != nil {
+		return nil, err
+	}
+	matchIdxs := varsRegex.FindAllSubmatchIndex([]byte(value), -1)
```

Comment:
This reminds me we need some rules on what chars are allowed in variable names. Should we restrict it?

Reply:
Should we support all of them? Or reduce it to a-z and 2-3 special chars?